### ISSUE TITLE
Idea: image generation

### ISSUE THREAD

**Issue by sirnewton01** [russross/blackfriday#445]
I think that it could be an interesting feature to allow the auto-conversion of code blocks for certain types of declarative image languages, such as graphviz. For example, the following code block could be transformed into a graph on output using the graphviz tool before generating the HTML. The HTML would link to the image.

```graphviz
 graph graphname {
     a -- b -- c;
     b -- d;
 }
```

It could be that code blocks would need to specify that this is the behaviour that they want, such as a processing instruction like this: ```graphviz[generate-image]

Alternatively, code blocks may not be the right syntax and so a new construct should be added.

There are also other examples of image generation through textual representations, such as this one written in Go: https://github.com/esimov/diagram

**Comment by shurcooL** [comment]
It should be very viable to implement something like this via a custom renderer that overrides the rendering for the [`CodeBlock`](https://godoc.org/gopkg.in/russross/blackfriday.v2#CodeBlock) node.

**Comment by sirnewton01** [comment]
True, I was going to experiment with the [mdtopdf](https://github.com/mandolyte/mdtopdf)in my own fork. I was thinking that there could be a construct in the markdown language itself and perhaps some core library support.

**Comment by lukesmurray** [comment]
Sorry I have a couple quesitons beacuse I think this is very interesting. Related is the implementation of this functionality in [Markdown Preview Enhanced](https://shd101wyy.github.io/markdown-preview-enhanced/#/code-chunk) Which use syntax like the following

this would show the output of the go code

    ```go {cmd}
    ... some go code
    ```

this would show the output of the go code executed using executable at the path

    ```go {cmd=/usr/bin/go}
    ... some go code
    ```
this shows the output of calling go with some arguments. 

    ```go {cmd=true args=["-v"]}
    ... some go
    ```

could this curly bracket notation be parsed through a custom renderer?

This also raises security concerns since the renderer is going to be executing arbitrary code in the document. Could a flag be passed to the custom renderer to confirm that arbitrary code is safe to execute?

**Comment by dmitshur** [comment]
> could this curly bracket notation be parsed through a custom renderer?

I don't see why not.

> This also raises security concerns since the renderer is going to be executing arbitrary code in the document.

Indeed. Care should be taken to avoid running it on untrusted input.

**Comment by sirnewton01** [comment]
When I wrote the issue initially, I wasn't anticipating that shell commands would be passed in. Instead, just predefined keywords that would tell the renderer that the graphviz or similar renderer should produce an image for the block using the source as the input. I don't think that this mode would require much in the way of additional security considerations unless there is some way to craft malicious graphviz. :)

Agreed that if arbitrary shell commands will be used that there will need to be additional security measures to prevent easy exploitation.

**Comment by lukesmurray** [comment]
Markdown Preview Enhanced actually uses your idea as well as arbitrary shell commands. The keyword `dot` can be used to create a graphviz diagram. More example can be found [here](https://shd101wyy.github.io/markdown-preview-enhanced/#/diagrams).
